README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# conflr

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/line/conflr.svg?branch=master)](https://travis-ci.org/line/conflr)
[![CRAN
status](https://www.r-pkg.org/badges/version/conflr)](https://CRAN.R-project.org/package=conflr)
<!-- badges: end -->

conflr is an R package to post [R
Markdown](https://rmarkdown.rstudio.com/) documents to
[Confluence](https://www.atlassian.com/software/confluence), a content
collaboration tool by Atlassian.

## Installation

Install conflr from CRAN with:

``` r
install.packages("conflr")
```

Alternatively, if you need the development version, install it with:

``` r
# install.packages("devtools")
devtools::install_github("line/conflr")
```

## Preparation

conflr uses these environmental variables to access your Confluence.

  - `CONFLUENCE_URL`: The base URL of your Confluence. e.g.
    `https://confluence.example.com` (On Atlassian Cloud,
    `https://<your-domain>.atlassian.net/wiki`).
  - `CONFLUENCE_USERNAME`: Your username (On Atlassian Cloud, your email
    address).
  - `CONFLUENCE_PASSWORD`: Your password (On Atlassian Cloud, your API
    token. For more details about API token, please read [the official
    document](https://confluence.atlassian.com/cloud/api-tokens-938839638.html)).

There are several ways to set these environmental variables. The
quickest way is to enter in the popups that are displayed when you run
the addin (see Usages section below). The inputs are cached in the
environmental variables listed above by default.

Another way is to set the variables in the `.Renviron` file (you can
open the file with `usethis::edit_r_environ()`). For example, you can
set the base URL in the file as the following.

    CONFLUENCE_URL=https://confluence.example.com

## Usages

conflr provides the following ways to post R Markdown documents to
Confluence.

1.  Use an RStudio Addin
2.  Run `confl_create_post_from_Rmd()` on console
3.  Specify `conflr::confluence_document` on the YAML
front-matter

### RStudio Addin

#### 1\. Move focus to the .Rmd file and click “Post to Confluence” Addin

(**Caution for those who are not familiar with R Markdown**: R
Markdown’s powerfulness allows you to execute arbitrary code; be sure
about what the code does before clicking “Post to Confluence”\!)

![](man/figures/screenshot1.png)

![](man/figures/screenshot2.png)

Then, you will be asked your username and password.

![](man/figures/popup1.png)

#### 2\. Check the preview and click “Publish”

  - **type**: The type of the page (*page* means a normal wiki page,
    whereas *blogpost* mean a blog post, not a page).
  - **Space Key**: The key of the space you want to post.
  - **Parent page ID**: (optional): The ID of the parent page to the
    page.
  - **Use original image sizes**: If checked, do not resize the images.
  - **Fold code blocks**: If checked, code blocks are folded by default.
  - **TOC**: If checked, add a Table of Contents.
  - **TOC depth**: The lowest heading level to include in the Table of
    Contents.

![](man/figures/screenshot3.png)

#### 3\. Check the result

![](man/figures/screenshot4.png)

### `confl_create_post_from_Rmd()`

If you don’t use RStudio, you can use `confl_create_post_from_Rmd()`.
The basic usage is

``` r
confl_create_post_from_Rmd("~/path/to/your.Rmd")
```

#### Batch use

If you want to use this function without interaction, specify
`interactive = FALSE`. This skips any confirmations or previews.

``` r
confl_create_post_from_Rmd("~/path/to/your.Rmd", interactive = FALSE)
```

Note that, if you want to run this periodically, you also need to set
`update = TRUE` to allow conflr to overwrite the existing
page.

``` r
confl_create_post_from_Rmd("~/path/to/your.Rmd", interactive = FALSE, update = TRUE)
```

### `conflr::conflence_document`

conflr’s functionality is also available as a custom R Markdown format;
You can specify `conflr::confluence_document` to `output` in the front
matter of your R Markdown document.

For example, if you set the following front matter, pressing `Knit`
button on RStudio (or running `rmarkdown::render()`) will publish the R
Markdown document to Confluence.

``` md
---
title: "title1"
output:
  conflr::confluence_document:
    space_key: "space1"
    update: true
---

...
```

For the detail about available options, please refer to
[`?confluence_document`](https://line.github.io/conflr/reference/confluence_document.html).

## Options

conflr recognizes these options:

  - `conflr_supported_syntax_highlighting`: A character vector of
    languages that your Confluence supports the syntax highlighting in
    addition to the default languages (by default, `sql`, `cpp`,
    `python`, `html`, `css`, `bash`, and `yaml` are supported).
  - `conflr_addin_clear_password_after_success`: If `TRUE`, unset
    `CONFLUENCE_PASSWORD` after the page is successfully uploaded via
    addin.

## Know limitations

### LaTeX support

conflr supports [Math
expressions](https://bookdown.org/yihui/rmarkdown/markdown-syntax.html#math-expressions)
to some extent. But, it requires [LaTeX Math
addon](https://marketplace.atlassian.com/apps/1210882/latex-math)
installed. Otherwise, you will see “unknown macro” errors on the page.

### htmlwidgets (e.g. leaflet, plotly)

conflr doesn’t support htmlwidgets-based packages like leaflet and
plotly. Instead, you can embed the screenshot by setting
`screenshot.force = TRUE` in the chunk option (c.f.
<https://bookdown.org/yihui/bookdown/html-widgets.html>).

## Advanced Usages

conflr is also a (non-complete) binding to [Confluence’s REST
API](https://developer.atlassian.com/server/confluence/confluence-server-rest-api/).
These low-level functions might be useful when you need to access to
your Confluence programmatically.

``` r
library(conflr)

# list pages
res <- confl_list_pages(spaceKey = "foo")
purrr::map_chr(res$results, "id")

# get page info
page <- confl_get_page(res$results[[2]]$id)
page$title

# create a page
new_page <- confl_post_page(
  spaceKey = "foo",
  title = "Test",
  body = glue::glue(
    '<ac:structured-macro ac:name="code">
     <ac:plain-text-body><![CDATA[this is my code]]></ac:plain-text-body>
     </ac:structured-macro>
    '))
new_page$`_links`
```

## How to contribute

See [CONTRIBUTING.md](https://github.com/line/conflr/blob/master/CONTRIBUTING.md)

## License

    Copyright (C) 2019 LINE Corporation
    
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, version 3.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

See [LICENSE.md](https://github.com/line/conflr/blob/master/LICENSE.md) for more detail.


NEWS.md
# conflr (development version)

* New `minor_edit` parameter (default to `FALSE`) in the YAML front
  matter to skip sending notifications on page or attachment updates
  to watchers (#20, #107 @daroczig, #108).

* New helper functions with the `confl_macro` prefix to generate HTML
  tags for the following Confluence macros: Table of Contents, Jira
  ticket references, Expand and Excerpt blocks (#111 @daroczig).

* Add an experimental support for tabset (#113).

# conflr 0.1.1

* A maintainance relase to fix errors on CRAN check.

# conflr 0.1.0

* First CRAN release.

## Major changes

* conflr now works outside RStudio (e.g. Emacs/Vim) (#10).

* conflr now fits batch/console uses; you can either
    1. run `confl_create_post_from_Rmd()` with `interactive = FALSE` so that it
       doesn't show Shiny popups (#32, @ndiquattro).
    2. set `output: conflr::confluence_document` in the YAML front matter of the
       R Markdown file, and run `rmarkdown::render()` (#44, @kazutan / #80).

    ``` md
    ---
    title: "I love Confluence"
    output:
      conflr::confluence_document:
        space_key: "space1"
    ---
    ```

* conflr provides several new options:
    * "Use original image sizes" option controls whether to resize the image
      (default) or not (#21).
    * "TOC" option adds a table of contents and "TOC depth" option changes the
      max level of headers to include (#67).
    * "Fold code blocks" option controls whether to fold codes (default) or not
      (#81).

### Minor changes

* `confl_create_post_from_Rmd()` gets `params` argument for parameterized R
  Markdown (#37, @ellisvalentiner).

* External images are now converted properly (#39).

* Add an option `conflr_addin_clear_password_after_success` not to cache the
  password as envvars. (#41 and #48, @Curycu).

* A new function `confl_contentbody_convert()` converts the Confluence-related
  formats by using the Confluence REST API (#58).

* `confl_post_page()` and `confl_update_page()` no longer translate the
  Confluence macros automatically. Accordingly, they lose `image_size_default`
  and `supported_syntax_highlighting` arguments (#76).

* `confl_create_post_from_Rmd()` now handles documents containing Confluence
  macro tags (i.e. `<ac:...>` or `<ri:...>`) properly (#76).

# conflr 0.0.5

* Initial release on GitHub


R/addin.R
# Copyright (C) 2019 LINE Corporation
#
# conflr is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, version 3.
#
# conflr is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See <http://www.gnu.org/licenses/> for more details.

#' @param Rmd_file
#'   Path to an .Rmd file.
#' @param params
#'   If provided, a list of named parameters that override custom
#'   params in the YAML front-matter.
#' @param ...
#'   Arguments passed to `confluence_documents()`.
#'
#' @return
#'   `confl_create_post_from_Rmd()` returns the URL of the published page.
#'
#' @rdname confluence_document
#'
#' @examples
#' example_Rmd <- system.file("extdata/example.Rmd", package = "conflr")
#'
#' \dontrun{
#' # Convert an R Markdown document into a 'Confluence' page interactively
#' confl_create_post_from_Rmd(example_Rmd)
#'
#' # You can override most of the parameters of confluence_document()
#' confl_create_post_from_Rmd(example_Rmd, space = "space1", toc = TRUE)
#' }
#'
#' @export
confl_create_post_from_Rmd <- function(Rmd_file, interactive = NULL, params = NULL, ...) {

  # FIXME: uncomment this when r-lib/ellipsis#32 is fixed
  # ellipsis::check_dots_used()

  if (is.null(interactive)) {
    interactive <- interactive()
  }

  # Sanity checks -----------------------------------------------------------

  if (!tolower(tools::file_ext(Rmd_file)) %in% c("rmd", "rmarkdown")) {
    abort(glue("{basename(Rmd_file)} is not .Rmd file!"))
  }

  # Combine options from arguments and front matter -------------------------

  # Get options on the front matter
  front_matter <- rmarkdown::yaml_front_matter(Rmd_file)
  if (is.list(front_matter$output) &&
    has_name(front_matter$output, "conflr::confluence_document")) {
    output_options <- front_matter$output$`conflr::confluence_document`
  } else {
    output_options <- list()
  }

  # Override the options by those via arguments (title will be handled in post_processor())
  output_options <- purrr::list_modify(output_options, ..., interactive = interactive)

  output_format <- exec(confluence_document, !!!output_options)

  # Knit --------------------------------------------------------------------

  output_file <- rmarkdown::render(
    input = Rmd_file,
    output_format = output_format,
    encoding = "UTF-8",
    params = params,
    # TODO: I'm not fully sure the global env is always the right place to knit, but this is needed to avoid
    #       an error related to data.table (#29). If this doesn't work, I need to add this code (c.f. https://github.com/Rdatatable/data.table/blob/5ceda0f383f91b7503d4a236ee4e7438724340be/R/cedta.R#L13):
    #   assignInNamespace("cedta.pkgEvalsUserCode", c(data.table:::cedta.pkgEvalsUserCode, "conflr"), "data.table")
    env = globalenv()
  )

  # Read the result URL (ignore warning of "incomplete final line")
  readLines(paste0(output_file, "_result_url"), warn = FALSE)
}

confl_create_post_from_Rmd_addin <- function() {
  if (!rstudioapi::isAvailable()) {
    abort("This function must be called on RStudio!")
  }

  Rmd_file <- rstudioapi::getSourceEditorContext()$path
  if (identical(Rmd_file, "")) {
    # Probably "UntitledX"
    abort("Please save the .Rmd file first!")
  }

  confl_create_post_from_Rmd(Rmd_file, interactive = TRUE)
}

confl_upload_interactively <- function(title, html_text, imgs, imgs_realpath,
                                       space_key = NULL,
                                       parent_id = NULL,
                                       type = "page",
                                       toc = FALSE, toc_depth = 7,
                                       code_folding = "none",
                                       supported_syntax_highlighting = getOption("conflr_supported_syntax_highlighting"),
                                       use_original_size = FALSE,
                                       minor_edit = FALSE) {

  # This will be assigned in Shiny's server function
  result_url <- NULL

  # Shiny UI -----------------------------------------------------------
  ui <- confl_addin_ui(
    title = title,
    html_text = html_text,
    imgs = imgs,
    imgs_realpath = imgs_realpath,
    space_key = space_key,
    parent_id = parent_id,
    type = type,
    toc = toc,
    toc_depth = toc_depth,
    code_folding = code_folding,
    minor_edit = minor_edit,
    use_original_size = use_original_size
  )


  # Shiny Server -------------------------------------------------------
  server <- function(input, output, session) {
    id <- NULL
    done <- shiny::reactiveVal(FALSE)

    shiny::observeEvent(input$confirm, {
      # if the space key is empty, tell the user to provide it and exit early
      if (identical(input$space_key, "")) {
        shiny::showModal(
          shiny::modalDialog("Please provide a space key!",
            footer = shiny::modalButton("OK")
          )
        )
        return()
      }

      id <<- try_get_existing_page_id(title = title, space_key = input$space_key)

      # If there is already an existing page, confirm user
      if (!is.null(id)) {
        shiny::showModal(shiny::modalDialog(
          glue(
            "There is already an existing page named '{title}'.\n",
            "Are you sure to overwrite it?"
          ),
          footer = shiny::tagList(
            shiny::modalButton("Cancel"),
            shiny::actionButton("trigger_done", "OK")
          )
        ))
      } else {
        done(TRUE)
      }
    })

    shiny::observeEvent(input$trigger_done, {
      shiny::removeModal()
      done(TRUE)
    })

    shiny::observe({
      shiny::req(done())

      result_url <<- confl_upload(
        title = title,
        html_text = html_text,
        imgs = imgs,
        imgs_realpath = imgs_realpath,
        id = id,
        space_key = input$space_key,
        parent_id = input$parent_id,
        type = input$type,
        session = session,
        toc = input$toc,
        toc_depth = input$toc_depth,
        code_folding = if (input$code_folding) "hide" else "none",
        supported_syntax_highlighting = supported_syntax_highlighting,
        use_original_size = input$use_original_size,
        minor_edit = minor_edit,
        # Already confirmed
        update = TRUE
      )

      unset_password_if_special_envvar_is_set()
    })
  }

  viewer <- shiny::dialogViewer("Preview", width = 1000, height = 800)
  shiny::runGadget(ui, server, viewer = viewer)

  result_url
}

# if the user doesn't want to store the password as envvar, clear it.
unset_password_if_special_envvar_is_set <- function() {
  if (isTRUE(getOption("conflr_addin_clear_password_after_success"))) {
    message("unsetting CONFLUENCE_PASSWORD...")
    Sys.unsetenv("CONFLUENCE_PASSWORD")
  }
}

read_utf8 <- function(x) {
  paste(readLines(x, encoding = "UTF-8"), collapse = "\n")
}

# c.f. https://kevinushey.github.io/blog/2018/02/21/string-encoding-and-r/
write_utf8 <- function(x, f) {
  # Ensure the text is encoded as UTF-8
  x <- enc2utf8(x)

  con <- withr::local_connection(file(f, open = "w+", encoding = "native.enc"))
  writeLines(x, con = con, useBytes = TRUE)
}

extract_image_paths <- function(html_text) {
  html_doc <- xml2::read_html(html_text)
  img_nodes <- xml2::xml_find_all(html_doc, ".//img")
  img_paths <- xml2::xml_attr(img_nodes, "src")
  # exclude external images
  img_paths[!is.na(img_paths) & !grepl("^https?://", img_paths)]
}

try_get_personal_space_key <- function(username) {
  # check if the space really exists
  tryCatch(
    confl_get_space(spaceKey = paste0("~", username))$key,
    error = function(e) {
      # Do not show even warnings because it's likely to happen as the keys of personal spaces are often numeric (#30).
      NULL
    }
  )
}

wrap_with_column <- function(..., width = 2) {
  shiny::column(width = width, ...)
}

# NOTE: conflr_supported_syntax_highlighting cannot be set via GUI because
#       it's not a feature frequently used and is a bit difficult to input
#       via Shiny interface.
confl_addin_ui <- function(title, html_text, imgs, imgs_realpath,
                           space_key = NULL, parent_id = NULL, type = "page",
                           toc = FALSE, toc_depth = 7,
                           code_folding = "none",
                           use_original_size = FALSE,
                           minor_edit = FALSE) {
  # title bar
  title_bar_button <- miniUI::miniTitleBarButton("confirm", "Publish", primary = TRUE)
  title_bar <- miniUI::gadgetTitleBar("Preview", right = title_bar_button)

  # type (page or blogpost)
  type_input <- shiny::selectInput(inputId = "type", label = "Type", choices = c("page", "blogpost"), selected = type)

  # space_key
  space_key_input <- shiny::textInput(inputId = "space_key", label = "Space Key", value = space_key)

  # parent page ID
  parent_id_input <- shiny::textInput(inputId = "parent_id", label = "Parent page ID", value = parent_id)

  # use the original size or not
  use_original_size_input <- shiny::checkboxInput(inputId = "use_original_size", label = "Use original image sizes", value = use_original_size)

  minor_edit_input <- shiny::checkboxInput(inputId = "minor_edit", label = "Minor edit", value = minor_edit)

  # add TOC or not
  toc_input <- shiny::checkboxInput(inputId = "toc", label = "TOC", value = toc)
  toc_depth_input <- shiny::numericInput(
    inputId = "toc_depth",
    label = "TOC depth",
    value = toc_depth,
    width = "4em"
  )
  code_folding_input <- shiny::checkboxInput(
    inputId = "code_folding",
    label = "Fold code blocks",
    value = identical(code_folding, "hide")
  )

  # Preview
  html_text_for_preview <- embed_images(html_text, imgs, imgs_realpath)
  preview_html <- shiny::HTML(html_text_for_preview)

  miniUI::miniPage(
    title_bar,
    miniUI::miniContentPanel(
      shiny::fluidRow(
        wrap_with_column(type_input),
        wrap_with_column(space_key_input),
        wrap_with_column(parent_id_input),
        wrap_with_column(
          use_original_size_input,
          code_folding_input,
          minor_edit_input
        ),
        wrap_with_column(
          toc_input,
          toc_depth_input
        )
      ),
      shiny::hr(),
      shiny::h1(title, align = "center"),
      shiny::div(preview_html)
    )
  )
}
